libfwupdplugin/fu-device.c

```diff
@@ -2972,6 +2972,63 @@ fu_device_probe_invalidate (FuDevice *self)
 	priv->done_setup = FALSE;
 }
 
+/**
+ * fu_device_report_metadata_pre:
+ * @self: A #FuDevice
+ *
+ * Collects metadata that would be useful for debugging a failed update report.
+ *
+ * Returns: (transfer full): A #GHashTable
+ *
+ * Since: 1.5.0
+ **/
+GHashTable *
+fu_device_report_metadata_pre (FuDevice *self)
+{
+	FuDeviceClass *klass = FU_DEVICE_GET_CLASS (self);
+	g_autoptr(GHashTable) metadata = NULL;
+
+	g_return_val_if_fail (FU_IS_DEVICE (self), NULL);
+
+	/* metadata for all devices */
+	metadata = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+	g_hash_table_insert (metadata,
+			     g_strdup ("GType"),
+			     g_strdup (G_OBJECT_TYPE_NAME (self)));
+
+	/* optionally subclassed */
+	if (klass->report_metadata_pre != NULL)
+		klass->report_metadata_pre (self, metadata);
+	return g_steal_pointer (&metadata);
+}
+
+/**
+ * fu_device_report_metadata_post:
+ * @self: A #FuDevice
+ *
+ * Collects metadata that would be useful for debugging a failed update report.
+ *
+ * Returns: (transfer full): A #GHashTable
+ *
+ * Since: 1.5.0
+ **/
+GHashTable *
+fu_device_report_metadata_post (FuDevice *self)
+{
+	FuDeviceClass *klass = FU_DEVICE_GET_CLASS (self);
+	g_autoptr(GHashTable) metadata = NULL;
+
+	g_return_val_if_fail (FU_IS_DEVICE (self), NULL);
+
+	/* metadata for all devices */
+	metadata = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+	/* optionally subclassed */
+	if (klass->report_metadata_post != NULL)
+		klass->report_metadata_post (self, metadata);
+	return g_steal_pointer (&metadata);
```

Comment:
Most device types won't have this new vfunc, I wonder if it's really worth the extra memory to create the hash table and return an empty hash table in that common case.  Instead why not

```
if (klass->report_metadata_post == NULL)
        return NULL;
metadata = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
klass->report_metadata_post (self, metadata);
return g_steal_pointer (&metadata);
```

Reply:
Sure, but this isn't fast path at all, so I went for clarity. It's called just once at startup if it's the first boot after a UEFI update.

Reply:
I guess if `GType` isn't always added then my arguement is null and void. Will fix.